src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/TypeMapping.cs

```diff
@@ -0,0 +1,95 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using System.Threading;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Entry type for interop type mapping logic.
+    /// </summary>
+    public static class TypeMapping
+    {
+#if !NATIVEAOT
+        private static Lock s_lock = new Lock();
+        private static Dictionary<Type, TypeMapLazyDictionary> s_externalTypeMapsByGroup = new Dictionary<Type, TypeMapLazyDictionary>();
+
+        [RequiresUnreferencedCode("Lazy TypeMap isn't supported in the Trimmer")]
+        private static TypeMapLazyDictionary CreateOrGet(RuntimeType typeGroup)
+        {
+            Debug.Assert(!typeGroup.IsGenericType, "Type group should not be generic");
```

Comment:
Relates to matching logic in native code. I'd like to understand what limits we should place on the types we support.

Reply:
I've made this "just work".